### PR TITLE
Document how to declare new modules

### PR DESCRIPTION
**Pull request by blinkseb**
This starts the documentation on how to add new modules. Only the basic stuff is introduced for the moment, which should already help users to get started if they want to add something. The documentation is only valid if https://github.com/MoMEMta/MoMEMta/pull/151 is merged (the only impacted part of this documentation is the `Sticky` part, everything else is valid if we don't merge the PR).

I plan to document advanced usage too (inputs and outputs, etc.), not sure when, I'll try today! In the mean time, don't hesitate if you have feedback on what it's already written :smiley:

### CODE REVIEW

File: docs/in-depth/new-module.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+```C++
+REGISTER_MODULE(HelloWorldModule)
+    .Sticky();
+```
+
+Because your module does not produce any output, MoMEMta will automatically remove it from the computation graph, since it does not contribute in any way to the final integrand value. You can prevent this behaviour by registering your module as `sticky`: a `sticky` module is not removed from the computation graph if it's found to not contribute.
+
+## Loading the module
+
+So, you have successfully defining a new module and registered it with the system. In order to use it during the computation of the weight, you first have to _load_ the shared library where your module is.
```
Comment (swertz) [inline docs/in-depth/new-module.md]: defined

Comment (swertz) [reply docs/in-depth/new-module.md]: Would it be obvious to users how to compile the module into a library?

Comment (blinkseb) [reply docs/in-depth/new-module.md]: Not really I know, but that fell more into "How i compile C++" than "How I use MoMEMta", I'm not sure how to document that

**Comment by swertz** [comment]
Thanks a lot :+1:

We can wait a bit before merging so if you have more knock yourself out ;)

Maybe this page could be updated (top) to link the new page: https://github.com/MoMEMta/momemta.github.io/blob/website/docs/in-depth/calling-momemta.md

File: docs/in-depth/new-module.md (hunk 1 above)
Comment (swertz) [reply docs/in-depth/new-module.md]: Actually we could imagine having a script taking care of it?

Comment (blinkseb) [reply docs/in-depth/new-module.md]: Hm I'm not sure, it really depends how the user wants to build its code. You can use plain makefiles, cmake, autoconf, etc.

File: docs/in-depth/new-module.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+private: Value
+
+<double> input;
+};</double>
+
+````
+
+The first addition grab an input of type `double` from the memory pool, expected to be produced by a module named `module` and called `output` (remember that the _output_ of a module becomes the _input_ of another module). The second addition print the actual value of the input at each integration step. In order to access the value of the input, you must dereference the variable (`Value` actually acts as a pointer to the real value, so you can also use the arrow (`->`) operator).
```
Comment (swertz) [inline docs/in-depth/new-module.md]: `grabs` and `prints`

File: docs/in-depth/new-module.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+private:
+  Value<double> input;
+
+  std::shared_ptr<double> output;
+};
+```
+
+First addition register an output named `my_output` for this module, of type `double`. The second addition set the value of the output to be twice the input value.
```
Comment (swertz) [inline docs/in-depth/new-module.md]: same here ;)

File: docs/in-depth/new-module.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+In the above example, we hardcoded the `InputTag` used to retrieve the input. This is not a good practice, because you can't know when writing your module how the user will want to use your module and which output he wants to connect to your module. To overcome this, you can use the parameters (second argument of the constructor) to get data specified by the user inside the configuration file. In our example here, we can request the user to tell, inside the configuration, how he wants to connect our module by letting him specify the `InputTag` to use.
+
+Parameters are stored into a [`ParameterSet`](https://momemta.github.io/MoMEMta/dev/classParameterSet.html), which is simply a dictionary of heterogeneous types. You can access values using the `get` function, specifying the name of the parameter you want.
+
+!!! note
+    Only a subset of types are supported by `ParameterSet`. You can only store and retrieve `int64_t` (integers), `double` (floating point numbers), `bool`, `std::string`, `InputTag`, `ParameterSet`, as well as `std::vector` of any of the previous types.
+
+We will modify our module to remove the hardcoded `InputTag` and except the user to pass it as a parameter:
```
Comment (swertz) [inline docs/in-depth/new-module.md]: `except` -> `require`?

**Comment by swertz** [comment]
Thanks a lot for the additions! :+1: It's looking really good, I think there is enough for now or do you have something else in mind?! 
The only thing that's perhaps less clear is the difference of `[Optional]Input` with and without 's'?

**Comment by blinkseb** [comment]
I think I documented everything I wanted :) Indeed the `OptionalInput(s)` is a bit tricky, maybe it's better to just document the `Input` method? I was kind of hopping the naming would be self-explanatory but:
  - `Input` expect a single `InputTag`
  - `Inputs` expect a vector of `InputTag`
  - `Optional` mark the input as... optional :smile:

File: docs/in-depth/new-module.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+```c++
+REGISTER_MODULE(HelloWorldModule)
+    .Attr("mass: double")
+    .Input("which_input")
+    .Output("output");
+```
+
+##### Supported types
+
+The following types are supported when registering a new attribute:
```
Comment (swertz) [inline docs/in-depth/new-module.md:227]: There's also the Paths but maybe that's too technical to mention here?

**Comment by swertz** [comment]
Yes it's probably fine; if a user wants to write a module he'll probably end up checking how it's done for  other modules anyway ;)

**Comment by blinkseb** [comment]
Indeed that would make sense! I think I'll add a sentence at the end with a link to the `modules` directory of the MoMEMta repo to help the users browse available modules.

File: docs/in-depth/new-module.md (hunk 5 above)
Comment (blinkseb) [reply docs/in-depth/new-module.md:227]: Indeed you're right, but Paths are not intended to be used by others modules than the Looper (it won't work at all) so let's keep that under the hood for the moment 😄

**Comment by blinkseb** [comment]
I updated the doc, removing the `Inputs` and `Optional...` description. I also added at the end a link to the modules source code.

On my side I think it's ready, but as I said, it's only valid if the associate PR In MoMEMta is merged, so we can probably wait a bit before merging this one in case there're changes to be done

**Comment by swertz** [comment]
Thanks! I agree, we can wait some more, I'm still going through the PR...

**Comment by swertz** [comment]
This can now be merged since MoMEMta has been updated. Thanks again! :+1:

**Pull request merged**